### ISSUE TITLE
Multi-tenant design for secure app that include member users and multiple organizations

### ISSUE THREAD

**Issue by marianolc** [aws-amplify/amplify-cli#5102]
**Which Category is your question related to?**

Cognito, GraphQL, DynamoDB

**Amplify CLI Version** 

4.24.1

**What AWS Services are you utilizing?**

AppSync, Cognito, DynamoDB

**Provide additional details e.g. code snippets**

I'm trying to design a multi-tenant app using DataStore and AppSync. The app needs to follow the following rules:

1) Should allow free registration of an user account
2) Should allow the user to register an organization (a tenant). The user who registers an organization would become its admin
3) Should allow an admin to create staff records for his organization
4) Should allow admin to send invites to other registered users and upon verification that user would be registered as the staff with the specified role (admin or member)
5) Should allow a user to be a staff of multiple organizations
6) Should allow staff of an organization to manage (create, read, update & delete) they records.
7) Should allow admin to read all the data of all his staff
8) Should not allow user with member role to read/write records of other members of his and/or other organizations
9) Should not allow admin of an organization to read/write records of other organizations where he is not registered and/or is a member

_those rules are extracted from another blogpost for another engine/service._

(1) is achieved using cognito. This is no issue.

(2) This is achieved using the @auth directive with the owner field:

```
type Organization @model 
  @auth(rules: [{allow:owner}]) 
  @key(fields: ["owner","name"]) 
{
  id: ID!
  owner: ID!
  name: ID!
  description: String
}
```

_Note that i'm using owner+name as Partition Key and Sort Key with an extra ID field because DataStore needs that field to exists locally. So its basically using ID as local PK and owner+name as remote._

(4) Here is were the trouble begins. I know that amplify allow to add user to a group using code. But should the organizations become a cognito group to achieve this?

(6) This is the other big issue. This is basically row-level access for every user based on the organization(or group?) he belongs.

How can I achieve this kind of security access using amplify?

**Comment by marianolc** [comment]
After some thinking, I came to some Ideas I would like to know if this is the best way to achieve it:

(4) The admin will simply add records to a new table that joins userId with Organization. 

(5) This will be a non issue with the model defined in (4)

(6) To achieve this, I think I will need to create a lambda for "Cognito Pre-Token Generation trigger" like this one:

```
exports.handler = (event, context, callback) => {
    event.response = {
        "claimsOverrideDetails": {
            "claimsToAddOrOverride": {
                // replace this with DynamoDB data
                "organizations": "org1;org2"
            }
        }
    };

    // Return to Amazon Cognito
    callback(null, event);
};
```

Now, for example If I want to create a table for products it would looks like this:

```
type Product @model 
  @key(fields: ["organization","sku"]) 
{
  id: ID!
  organization: ID!
  sku: ID!
  name: String
  price: Float  
}
```

Then somehow (still didn't figure it yet) modify the **Resolver for Query.listProducts** and force the use of one of those two organizations.

Is this the correct approach?

**Comment by RossWilliams** [comment]
You are on the right track, but a few tips.
1. You mention using DataStore. DataStore has some important limitations that make a multi-tenant app design harder. Be careful as some auth patterns do not work with DataStore.
2. A lot of these requirements are for write operations. Using a lambda resolver you can handle very complex write authorisation rules.
3. The docs have a really good set of access patterns [here](https://docs.amplify.aws/cli/graphql-transformer/dataaccess). They will give you a good starting point for your read operations.
4. You can override a resolver by creating a folder "amplify/backend/api/$API_NAME/resolvers, and add resolver templates in that folder with the exact name of the resolver you want to override.

**Comment by marianolc** [comment]
So with the resolvers I was able to "force" the organization the user belongs with something like this:

```
  #set( $organizations = $ctx.identity.claims.get("organizations"))

## change this with the correct organization list
#if($organizations!="org1;org2")
{
	$util.error($organizations, "InvalidArgumentsError")
}
#end
```

I think to this being able to work with datastore insted of forcing this parameter I will need to filter the results like it's done with  @owner

Now I think I have to solve what happens when the user is removed from the organization, the JWT will still have this value listed so he will still be able to query that data.  I'm thinking two ways of "solving" this:

1) Expiring the JWT token everytime the user is added(maybe for this case is not needed) or removed from an organization.
2) Insted of getting the organization list from the JWT just query directly the table where the relations are. Is this possible?

Any other solution i'm not thinking?

Thanks for your help!

**Comment by RossWilliams** [comment]
I don't think you will get DataStore and dynamic group auth to work well together, be careful investing much time going down this route.

Removing users from an organisation 

1a. You can't "expire the JWT token". You can configure Cognito to have short-lived access tokens (5 minute minimum?).
1b. You could also consider how often users leave an organisation and how important it is to immediately revoke access.

2a. Query directly, you can do this by routing your queries "through" the organisation table. If organisation had an "owners" field that is an array of usernames, then owner auth would stop working as soon as the user was removed from the organisation. (assumes you don't have 1000+ users in a org)
```graphql
query {
    getOrganisation(id: "1") {
       products {
           items {
               thing
           }
       } 
   }
}
```
2b. What Amplify doesn't support well right now is DynamoDB Transactions. If transactions were well supported, you could combine a Get on an item with a Check that a user belongs to an organisation at the same time. 2a is annoying because it slows down your API performance with 2 sequential requests. 2b runs the requests in parallel.

**Comment by edwardfoyle** [comment]
Hi @marianolc this would be a good question to ask the Amplify community in our Discord channel: [discord.gg/amplify](https://discord.gg/amplify)

Another thing you may want to look into is fine-grained access control policies for DynamoDB: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/specifying-conditions.html. Amplify CLI does not support / configure these out of the box, but you could potentially modify the policies created by Amplify to use this strategy
